### PR TITLE
EAD zeroconf: send full CRED_V in message_2

### PR DESCRIPTION
**Pull request by geonnave**
Included in this PR:
- [x] have Initiator and Responder constructors accept optional type for its counterpart (:warning: this changes the API!)
- [x] send full CRED_V in message_2
- [x] add a mocked W (only the needed parts to make the handshake with zero-touch work)

Sending CRED_R in message_3 would also make sense, but at this point we still don't need it.

### CODE REVIEW

File: lib/src/edhoc.rs [hunk 1]
```diff
@@ -483,60 +492,86 @@ pub fn i_process_message_2(
...
 
             if plaintext_2_decoded.is_ok() {
-                let (c_r_2, kid, mac_2, ead_2) = plaintext_2_decoded.unwrap();
+                let (c_r_2, id_cred, mac_2, ead_2) = plaintext_2_decoded.unwrap();
                 c_r = c_r_2;
 
                 // Step 3: If EAD is present make it available to the application
-                let ead_success = if let Some(ead_2) = ead_2 {
-                    i_process_ead_2(ead_2, cred_r_expected, &h_message_1).is_ok()
+                let (ead_ok, r_authenticated_via_ead, cred_r) = if let Some(ead_2) = ead_2 {
```
Comment (malishav) [inline lib/src/edhoc.rs:499]: If possible, rewrite to be more explicit, as discussed offline

File: lib/src/edhoc.rs [hunk 2]
```diff
@@ -483,60 +492,86 @@ pub fn i_process_message_2(
...
+                    // if EAD-zeroconf is present, then id_cred must contain a full credential
```
Comment (malishav) [inline lib/src/edhoc.rs:500]: Ideally, this should check if the EAD handler is the zero touch draft handler. No action needed at this point, let's track.

File: lib/src/lib.rs [hunk 3]
```diff
@@ -487,42 +413,28 @@ mod test {
...
-        let mut responder = EdhocResponder::new(
-            state_responder,
-            R,
-            G_I,
-            ID_CRED_I,
-            CRED_I,
-            ID_CRED_R,
-            CRED_R,
-        );
+        let mut responder = EdhocResponderState::new(state_responder, R, CRED_V_TV, Some(CRED_I));
```
Comment (malishav) [inline lib/src/lib.rs:434]: We currently instantiate the responder with `CRED_I` of the Initiator. This should ideally be `None` and the responder should fetch, for the time being through a mock up, the value of the `CRED_I` based on `ID_CRED_I`.

**Review by malishav — COMMENTED** [review]
Some comments inline, as discussed offline.

File: consts/src/lib.rs [hunk 4]
```diff
@@ -281,4 +303,231 @@ mod helpers {
 
         (info, info_len)
     }
+
+    pub fn parse_cred<'a>(cred: &'a [u8]) -> (BytesP256ElemLen, u8) {
+        let subject_len = (cred[2] - CBOR_MAJOR_TEXT_STRING) as usize;
+        let id_cred_offset: usize = 3 + subject_len + 8;
```
[the remark this answers is not in the document]
Comment (geonnave) [reply consts/src/lib.rs:309]: Addressed.

File: consts/src/lib.rs [hunk 5]
```diff
@@ -281,4 +303,231 @@ mod helpers {
...
+
+        (
+            cred[g_a_x_offset..g_a_x_offset + P256_ELEM_LEN]
+                .try_into()
+                .expect("Wrong key length"),
+            cred[id_cred_offset],
+        )
+    }
+
+    pub fn get_id_cred<'a>(cred: &'a [u8]) -> BytesIdCred {
```
[the remark this answers is not in the document]
Comment (geonnave) [reply consts/src/lib.rs:320]: For context: it is hardcoded and assumes that ID_CRED is a kid in a COSE header map.

File: consts/src/lib.rs [hunk 6]
```diff
@@ -281,4 +303,231 @@ mod helpers {
 
         (info, info_len)
     }
+
+    pub fn parse_cred<'a>(cred: &'a [u8]) -> (BytesP256ElemLen, u8) {
```
[the remark this answers is not in the document]
Comment (geonnave) [reply consts/src/lib.rs:307]: Addressed.

File: lib/src/edhoc.rs [hunk 7]
```diff
@@ -1257,40 +1112,55 @@ fn compute_mac_2(
...
+                2
+            } else if is_cbor_bstr_2bytes_prefix(plaintext_2[1])
+                && is_cbor_uint_2bytes(plaintext_2[2])
+                && (plaintext_2[3] as usize) < plaintext_2_len
+            {
+                let cred_len = plaintext_2[3] as usize;
+                id_cred_r = IdCred::FullCredential(&plaintext_2[4..4 + cred_len]);
+                4 + cred_len
+            } else {
+                0xff
```
[the remark this answers is not in the document]
Comment (geonnave) [reply lib/src/edhoc.rs:1137]: Addressed.

File: lib/src/edhoc.rs (hunk 1 above)
Comment (geonnave) [reply lib/src/edhoc.rs:499]: Tracked in #119 .

File: lib/src/edhoc.rs (hunk 2 above)
Comment (geonnave) [reply lib/src/edhoc.rs:500]: Tracked in #120.

File: lib/src/lib.rs (hunk 3 above)
Comment (geonnave) [reply lib/src/lib.rs:434]: Tracked in #121.

**Pull request merged**